plugins/upower/fu-plugin-upower.c

```diff
@@ -117,15 +117,14 @@ fu_plugin_update_prepare (FuPlugin *plugin,
 {
 	FuContext *ctx = fu_plugin_get_context (plugin);
 
-	/* not all devices need this */
-	if (!fu_device_has_flag (device, FWUPD_DEVICE_FLAG_REQUIRE_AC))
-		return TRUE;
 	if (flags & FWUPD_INSTALL_FLAG_IGNORE_POWER)
 		return TRUE;
 
 	/* not charging */
-	if (fu_context_get_battery_state (ctx) == FU_BATTERY_STATE_DISCHARGING ||
-	    fu_context_get_battery_state (ctx) == FU_BATTERY_STATE_EMPTY) {
+	if (fu_device_has_flag (device, FWUPD_DEVICE_FLAG_REQUIRE_AC) &&
+	    (fu_context_get_battery_state (ctx) ==
+		 FU_BATTERY_STATE_DISCHARGING ||
+	     fu_context_get_battery_state (ctx) == FU_BATTERY_STATE_EMPTY)) {
 		g_set_error_literal (error,
 				     FWUPD_ERROR,
 				     FWUPD_ERROR_AC_POWER_REQUIRED,
```

Comment:
do we want to move this into the engine perhaps?

Reply:
This again would require that powerd keeps the state up-to-date as it changes in the system via DBus signal handling and `fu_context_set_battery_state()`. At the moment powerd does not support said signal. @binaryNewts for additional details

Reply:
Ohh I think that's exactly what we want. It's much better to not allow the user to schedule the update than to "fail" after downloading and trying to deploy.

Reply:
I agree, but due to time constraints we may want to treat this as a future improvement :(

Reply:
Could you do a poll every minute using g_timer_add() and update the state/level then?

Reply:
@g2pop could you take a look at this and send a follow up change?

Reply:
Sure, I'll check it out.